File: C/diagsqt.c
#include "diagsqt.h"

// Function definitions.
// -----------------------------------------------------------------
// Compute X.^2*a and store the result in vector y.
void diagsqt (const MatrixElem* X, const double* a, double* y, 
	     Size m, Size n) {
  double t;  // An intermediate result.

  // Zero the entries of the result vector.
  setvector(y,0);

  // Repeat for each column of X.
  for (Index j = 0; j < n; j++, a++)

    // Repeat for each row of X.
    const double* yi = y;
    for (Index i = 0; i < m; i++, X++, yi++) {
      
      // Add X(i,j)^2 * a(j) to the ith entry of y.
      t    = (double) *X;
      *yi += t * t * (*a);
    }
}

File: MATLAB/diagsqt.m
% DIAGSQT(X) is the same as DIAG(X*X'), but the computation is done more
% efficiently, and without having to store an intermediate result of the
% same size as X. 
%
% DIAGSQT(X,A) efficiently computes DIAG(X*DIAG(A)*X').
function y = diagsqt (X, a)

  % If input A is not provided, set it to a vector of ones.
  [m n] = size(X);
  if ~exist('a')
    a = ones(n,1);
  end
  if length(a) ~= n
    error('Inputs X and A do not match');
  end

  % Convert A to a double column vector.
  a = double(a(:));
  
  % Convert X to single precision.
  if ~isa(X,'single')
    X = single(X);
  end
  
  % Compute the result using the efficient C routine.
  y = diagsqtmatlab(X,a);

  